core/server/common/src/main/java/alluxio/master/journal/AbstractJournalSystem.java

```diff
@@ -42,6 +56,60 @@ public synchronized void stop() throws InterruptedException, IOException {
     stopInternal();
   }
 
+  @Override
+  public void addJournalSink(Master master, JournalSink journalSink) {
+    mSinkLock.writeLock().lock();
+    try {
+      mJournalSinks.compute(master.getName(), (key, entry) -> {
+        if (entry == null) {
+          Set<JournalSink> s = new HashSet<>();
+          s.add(journalSink);
+          return s;
+        } else {
+          entry.add(journalSink);
+          return entry;
+        }
+      });
+      mAllJournalSinks.add(journalSink);
+    } finally {
+      mSinkLock.writeLock().unlock();
+    }
+  }
+
+  @Override
+  public void removeJournalSink(Master master, JournalSink journalSink) {
+    mSinkLock.writeLock().lock();
+    try {
+      mJournalSinks.computeIfPresent(master.getName(), (key, entry) -> {
+        if (entry != null) {
+          entry.remove(journalSink);
+          return entry;
+        }
+        return null;
+      });
+      // Compute the full set of sinks on removal.
```

Comment:
why not remove just `journalSink`?

Reply:
It is possible that a sink instance is associated with multiple masters, so removing it may not be correct.